File: java/medium/213_house_robber_ii.java
// https://leetcode.com/problems/house-robber

package java.medium;

class HouseRobbers {
    public int rob(int[] nums) {
        int n = nums.length;

        if (n == 0) {
            return 0;
        }

        if (n == 1) {
            return nums[0];
        }

        int rob1 = rob(nums, 0, n-2);
        int rob2 = rob(nums, 1, n-1);

        return Math.max(rob1, rob2);
    }

    public int rob(int[] nums, int start, int end) {
        int n = end - start + 1;

        int[] dp = new int[n];
        dp[0] = nums[start];
        if (n > 1) {
            dp[1] = Math.max(nums[start], nums[start+1]);
        }

        for (int i = 2; i < n; i++) {
            dp[i] = Math.max(dp[i-1], nums[start+i] + dp[i-2]);
        }

        return dp[n-1];
    }
}

File: go/medium/213_house_robber_ii.go
package medium

func RobII(nums []int) int {
	n := len(nums)
	if n == 1 {
		return nums[0]
	}
	sum1 := robIIFrom(nums, 0, n-2, make(map[int]int))
	sum2 := robIIFrom(nums, 1, n-1, make(map[int]int))

	return max(sum1, sum2)
}

func robIIFrom(nums []int, start, end int, memo map[int]int) int {
	if start > end {
		return 0
	}
	if _, ok := memo[start]; ok {
		return memo[start]
	}

	sum1 := nums[start] + robIIFrom(nums, start+2, end, memo)
	sum2 := robIIFrom(nums, start+1, end, memo)
	maxRob := max(sum1, sum2)

	memo[start] = maxRob

	return maxRob
}
